docs/architecture/003-handler-implementation.md

```diff
@@ -0,0 +1,552 @@
+# ADR 003: Handler implementation
+
+## Changelog
+* 2020-08-06: Initial proposal
+
+## Context
+
+> This section contains all the context one needs to understand the current state, and why there is a problem. It should be as succinct as possible and introduce the high level idea behind the solution.
+
+TODO
+
+## Decision
+
+In this chapter, we provide recommendation for implementing IBC handlers within the `ibc-rs` crate.
+Concepts are introduced in the order given by a topological sort of their dependencies on each other.
+
+### Events
+
+IBC handlers must be able to emit events which will then be broadcasted via the node's pub/sub mechanism,
+and eventually picked up by the IBC relayer.
+
+A generic interface for events is provided below, where an event is represented
+as a pair of an event type and a list of attributes. An attribute is simply a pair
+of a key and a value, both represented as strings. TODO: Describe event type.
+
+```rust
+pub struct Attribute {
+    key: String,
+    value: String,
+}
+
+impl Attribute {
+    pub fn new(key: String, value: String) -> Self;
+}
+
+pub enum EventType {
+    Message,
+    Custom(String),
+}
+
+pub struct Event {
+    tpe: EventType,
+    attributes: Vec<Attribute>,
+}
+
+impl Event {
+    pub fn new(tpe: EventType, attrs: Vec<(String, String)>) -> Self;
+}
+```
+
+### Logging
+
+IBC handlers must be able to log information for introspectability and ease of debugging.
```

Comment:
Not sure if introspectability is right (does that mean self-examination?). Maybe we mean here to say that the handlers log information for exporting it in a simple/plain format?